crates/driver/src/domain/competition/solution/mod.rs

```diff
@@ -270,31 +348,83 @@ impl std::fmt::Debug for Solution {
     }
 }
 
+/// Given two solutions returns a set of factors with
+/// which prices of the second solution would have to be multiplied so that the
+/// given token would have the same price in both solutions.
+fn scaling_factors(first: &Prices, second: &Prices) -> HashSet<BigRational> {
+    first
+        .keys()
+        .collect::<HashSet<_>>()
+        .intersection(&second.keys().collect::<HashSet<_>>())
+        .map(|&token| {
+            let first_price = first[token];
+            let second_price = second[token];
+            BigRational::new(
+                number::conversions::u256_to_big_int(&second_price),
+                number::conversions::u256_to_big_int(&first_price),
+            )
+        })
+        .collect()
+}
+
 /// Carries information how the score should be calculated.
 #[derive(Debug, Clone)]
 pub enum SolverScore {
     Solver(eth::U256),
     RiskAdjusted(f64),
 }
+
+impl SolverScore {
+    fn merge(&self, other: &SolverScore) -> Option<SolverScore> {
+        match (self, other) {
+            (SolverScore::Solver(a), SolverScore::Solver(b)) => Some(SolverScore::Solver(a + b)),
+            (SolverScore::RiskAdjusted(a), SolverScore::RiskAdjusted(b)) => {
+                Some(SolverScore::RiskAdjusted(a * b))
+            }
+            _ => None,
+        }
+    }
+}
+
 /// A unique solution ID. This ID is generated by the solver and only needs to
 /// be unique within a single round of competition. This ID is only important in
 /// the communication between the driver and the solver, and it is not used by
 /// the protocol.
-#[derive(Debug, Clone, Copy, Hash, PartialEq, Eq)]
-pub struct Id(pub u64);
+#[derive(Debug, Clone, Hash, PartialEq, Eq)]
+pub enum Id {
+    Single(u64),
+    Merged(Vec<Id>),
```

Comment:
Since the solutions get merged together anyway is there any value in seeing which solution got merged with which other in particular?
If not then `Merged(Vec<u64>)` seems more appropriate to me. (suggestion applies to all the other Ids in this PR as well)